File: 583-delete-operation-for-two-strings/583-delete-operation-for-two-strings.java
class Solution {
    public int minDistance(String word1, String word2) {
        int[][] memo = new int[word1.length()][word2.length()];
        for(int i =0;i<word1.length();i++){
            Arrays.fill(memo[i],-1);
        }
        
        return word1.length()+word2.length()-2*(recur(word1,word2,word1.length(),word2.length(),memo));
    }
    
    
    public int recur(String s1, String s2, int i,int j,int[][] memo){
        if(i==0 || j==0){
            return 0;
        }
        if(memo[i-1][j-1]!=-1){
            return memo[i-1][j-1];
        }
        
        if(s1.charAt(i-1)==s2.charAt(j-1)){
            memo[i-1][j-1] = 1+recur(s1,s2,i-1,j-1,memo);
        }
        else{
            memo[i-1][j-1] = Math.max(recur(s1,s2,i-1,j,memo),recur(s1,s2,i,j-1,memo));
        }
        return memo[i-1][j-1];
        
    }
}

File: 583-delete-operation-for-two-strings/583-delete-operation-for-two-strings.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        
        def recur(i1,i2,):
            if(i1<0 or i2<0):
                return 0
            if((i1,i2) in cache):
                return cache[(i1,i2)]
            if(word1[i1]==word2[i2]):
                cache[(i1,i2)] = 1+recur(i1-1,i2-1)
                return cache[(i1,i2)]
            p1 = recur(i1,i2-1)
            p2 = recur(i1-1,i2)
            cache[(i1,i2)] = max(p1,p2)
            return max(p1,p2)
        
        cache = dict()
        common = recur(len(word1)-1,len(word2)-1)
        w1_del = len(word1)-common
        w2_del = len(word2)-common
        return w1_del+w2_del
        
        
        